File: example/MyExtension/ShareViewController.m
//
//  ShareViewController.m
//  Ez File Transfer Pro Share
//
//  Created by ezerway on 30/06/2023.
//

#import "ShareViewController.h"
#import "MobileCoreServices/MobileCoreServices.h"

@interface ShareViewController ()

@end

@implementation ShareViewController

- (BOOL)isContentValid {
    // Do validation of contentText and/or NSExtensionContext attachments here
    return YES;
}

- (void)didSelectPost {
    // This is called after the user selects Post. Do the upload of contentText and/or NSExtensionContext attachments.
    
    // Inform the host that we're done, so it un-blocks its UI. Note: Alternatively you could call super's -didSelectPost, which will similarly complete the extension context.
    if ([self.extensionContext inputItems].count < 1) {
        [self.extensionContext completeRequestReturningItems:@[] completionHandler:nil];
        return;
    }
    
    NSExtensionItem *content = [[self.extensionContext inputItems] firstObject];
    NSArray<NSItemProvider *> *contents = [content attachments];
    
    for (NSItemProvider *attachment in contents) {
        if ([attachment hasItemConformingToTypeIdentifier: (NSString *) kUTTypeImage] != NO) {
            [attachment loadItemForTypeIdentifier:(NSString *) kUTTypeImage options: nil completionHandler:^(__kindof id<NSSecureCoding>  _Nullable data, NSError * _Null_unspecified error) {
                NSURL *imageUrl = (NSURL *) data;
                NSString *sharedImageUrl = [self saveImage:imageUrl];
                [self redirectToHostApp: sharedImageUrl];
                [self.extensionContext completeRequestReturningItems:@[] completionHandler:nil];
            }];
        }
    }
}

- (NSArray *)configurationItems {
    // To add configuration options via table cells at the bottom of the sheet, return an array of SLComposeSheetConfigurationItem here.
    return @[];
}

- (NSString *) saveImage: (NSURL *) url {
    NSData *imageData = [NSData dataWithContentsOfURL:url];
    UIImage *selectedImage = [UIImage imageWithData:imageData];

    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSURL *sharedContainer = [fileManager containerURLForSecurityApplicationGroupIdentifier:@"APP_GROUPS"];
    NSString *dirPath = sharedContainer.path;
    NSString *sharedImageFilePath = [NSString stringWithFormat:@"%@/%@", dirPath, @"image.png"];
    NSData *binaryImageData = UIImagePNGRepresentation(selectedImage);
    [binaryImageData writeToFile:sharedImageFilePath atomically:YES];
    return sharedImageFilePath;
}

- (void) redirectToHostApp: (NSString *) sharedUrl {
    NSString *appURL = [NSString stringWithFormat:@"SCHEMA://share?url=%@", sharedUrl];
    SEL selectorOpenURL = sel_registerName("openURL:");
    UIResponder *responder = (UIResponder *) self;
    
    while (responder != nil) {
        if ([responder respondsToSelector:(selectorOpenURL)] != FALSE) {
            [responder performSelector:(selectorOpenURL) withObject: [NSURL URLWithString: appURL]];
        }
        responder = responder.nextResponder;
    }
}

@end

File: _extensions/MyExtension/ShareViewController.swift
import UIKit
import Social
import MobileCoreServices

@objc(ShareViewController)
class ShareViewController: SLComposeServiceViewController {
    let hostAppURLScheme = "SCHEMA"
    let textImageType = kUTTypeImage as String

    override func isContentValid() -> Bool {
        // Do validation of contentText and/or NSExtensionContext attachments here
        return true
    }
    override func didSelectPost() {
      if let content = self.extensionContext!.inputItems[0] as? NSExtensionItem {
        if let contents = content.attachments {
          for (_, attachment) in (contents).enumerated() {
            if attachment.hasItemConformingToTypeIdentifier(self.textImageType) {
                attachment.loadItem(forTypeIdentifier: self.textImageType, options: nil) { data, error in
                    let imageUrl = data as NSURL;
                    let sharedImageUrl = self.saveImage(url: imageUrl)
                    self.redirectToHostApp(sharedURL: sharedImageUrl)
                    extensionContext!.completeRequest(returningItems: [], completionHandler: nil)
                }
            }
          }
        }
      }
    }

    override func configurationItems() -> [Any]! {
        // To add configuration options via table cells at the bottom of the sheet, return an array of SLComposeSheetConfigurationItem here.
        return []
    }

    func saveImage(url: NSURL) -> String {
        guard let imageData = NSData(contentsOfURL: url) else { return }
        let selectedImage = UIImage(data: imageData)

        let fileManager = NSFileManager.defaultManager()
        if let sharedContainer = fileManager.containerURLForSecurityApplicationGroupIdentifier("APP_GROUPS") {
            if let dirPath = sharedContainer.path {
                let sharedImageFilePath = dirPath + "/image.png"
                let binaryImageData = UIImagePNGRepresentation(selectedImage!)
                binaryImageData?.writeToFile(sharedImageFilePath, atomically: true)
                // send the sharedImageFilePath to the containing app
                return sharedImageFilePath;
            }
        }

        return "";
    }
  
    private func redirectToHostApp(sharedURL: String) {
          var urlComponents = URLComponents()
          urlComponents.scheme = hostAppURLScheme
          urlComponents.host = "share"
          urlComponents.path = "/"
          urlComponents.queryItems = [
              URLQueryItem(name: "url", value: sharedURL),
          ]
          // urlComponents.url: \(scheme)://share/?url=\(sharedURL)
          let url = urlComponents.url
          var responder = self as UIResponder?
          let selectorOpenURL = sel_registerName("openURL:")
          
          while (responder != nil) {
              if (responder?.responds(to: selectorOpenURL))! {
                  responder?.perform(selectorOpenURL, with: url)
              }
              responder = responder!.next
          }
      }
}
